test/e2e/create_workload_cluster_test.go

```diff
@@ -0,0 +1,110 @@
+// +build e2e
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+var _ = Describe("Workoad cluster creation", func() {
+	var (
+		ctx           = context.TODO()
+		specName      = "create-workload-cluster"
+		namespace     *corev1.Namespace
+		cancelWatches context.CancelFunc
+		cluster       *clusterv1.Cluster
+		clusterName   string
+	)
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		Expect(e2eConfig).ToNot(BeNil(), "Invalid argument. e2eConfig can't be nil when calling %s spec", specName)
+		Expect(clusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. clusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(bootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. bootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(artifactFolder, 0755)).To(Succeed(), "Invalid argument. artifactFolder can't be created for %s spec", specName)
+
+		Expect(e2eConfig.Variables).To(HaveKey(KubernetesVersion))
+		Expect(e2eConfig.Variables).To(HaveKey(CNIPath))
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, bootstrapClusterProxy, artifactFolder)
+
+		clusterName = fmt.Sprintf("capz-e2e-%s", util.RandomString(6))
+	})
+
+	AfterEach(func() {
+		cleanup(ctx, specName, bootstrapClusterProxy, artifactFolder, namespace, cancelWatches, cluster, e2eConfig.GetIntervals, skipCleanup)
+	})
+
+	Context("Create single controlplane cluster", func() {
+		It("Should create a single node cluster", func() {
+			cluster, _, _ = clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+				ClusterProxy: bootstrapClusterProxy,
+				ConfigCluster: clusterctl.ConfigClusterInput{
+					ClusterctlConfigPath:     clusterctlConfigPath,
+					KubeconfigPath:           bootstrapClusterProxy.GetKubeconfigPath(),
+					InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+					Flavor:                   clusterctl.DefaultFlavor,
```

Comment:
Does this fetch the flavor from the `templates/` folder or from the new template you added in e2e/data/infrastructure-azure/cluster-template.yaml? If it's the latter, why do we need to store additional templates just for E2E? this might lead to overhead for maintenance: if something is changed in the default template, we now need to remember to also change the e2e template. It would be nice if this could either 1) point to the existing templates or 2) auto-generate the e2e template in e2e/data. This is probably more of a framework question though and not really specific to this PR. 
cc @fabriziopandini

Reply:
I agree. I will update the PR and re-use our current default flavor at `./template/cluster-template.yaml`. In terms of using a specific flavor, you would have to define the favored YAML file in the [E2E configuration](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/28ede51fdaf6a95e878d6dff244660893232d7b4/test/e2e/config/azure-dev.yaml#L46-L47). For example, if you want to use machine pool cluster template, we would define it in the E2E configuration like:

```yaml
    files:
      - sourcePath: ../../../templates/cluster-template-machinepool.yaml
```

And in the test suite, we would have to specify the flavor by:
```go
...
Flavor: "machinepool",
...
```

Reply:
+1 to what @chewong said. The template location can be changed